File: frontend/src/app/model/application.dto.ts
export interface ApplicationDto {
  applicationId: string;
  jobId: string;
  candidateUserId: string;
  githubUrl: string;
  status: string;
  appliedAt: string;

  cvFileName: string;
  cvContentType: string;

  jobTitle?: string;
  candidateName?: string;
  candidateEmail?: string;
  jobFitScore?: number | null;
  requiredSkillsMatched?: string[];
  requiredSkillsMissing?: string[];
  experienceGap?: number | null;
  seniorityMatch?: boolean | null;
  embeddingScore?: number | null;
  skillScores?: {
    skill: string;
    score: number;           // effective score (post-qualifier-curve)
    status: 'matched' | 'partial' | 'missing';
    evidence?: string | null;
    reason?: string | null;
    // Qualifier-aware fields (BASIC / INTERMEDIATE / ADVANCED / EXPERT requirements):
    rawScore?: number;       // pre-curve raw score
    qualifier?: 'basic' | 'intermediate' | 'advanced' | 'expert' | 'any';
    qualifierBar?: number;   // 0 for 'any', else 45/65/80/90
    meetsQualifier?: boolean;
    gapFromQualifier?: number;
    signal?: 'strength' | 'meets' | 'gap' | 'critical_gap' | '';
  }[];
  requirementScores?: {
    category?: string | null;
    description?: string | null;
    score?: number | null;
    weight?: number | null;
    evidence?: string | null;
    skillLevel?: 'BASIC' | 'INTERMEDIATE' | 'ADVANCED' | 'EXPERT' | null;
    criticalGap?: boolean;
  }[];
  // Non-scoring advisory signals shown as banners on the application detail page.
  // Currently produced: 'distance_far' (ON_SITE job + candidate far from HQ),
  // 'name_mismatch' (CV / GitHub / LinkedIn names disagree).
  warnings?: {
    kind: 'distance_far' | 'name_mismatch' | string;
    severity: 'warning' | 'info' | string;
    message: string;
    details?: Record<string, unknown>;
  }[];
  strengths?: string[];
  weaknesses?: string[];
  recommendation?: string | null;
  interviewQuestions?: string[];
  scoreExplanation?: string | null;
}

File: backend/application-microservice/src/main/java/com/recrutment/application/dto/ApplicationDto.java
package com.recrutment.application.dto;

import com.recrutment.application.enums.ApplicationStatus;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.Instant;
import java.util.UUID;

@Data
@NoArgsConstructor
@AllArgsConstructor
public class ApplicationDto {
    private UUID applicationId;
    private UUID jobId;
    private String candidateUserId;
    private String githubUrl;
    private ApplicationStatus status;
    private Instant appliedAt;

    private String cvFileName;
    private String cvContentType;

    /** Set when an application is auto-withdrawn (e.g. candidate hired elsewhere). */
    private String withdrawalReason;

    private String jobTitle;
    private String candidateName;
    private String candidateEmail;
    private Integer jobFitScore;
    private java.util.List<String> requiredSkillsMatched;
    private java.util.List<String> requiredSkillsMissing;
    private Float experienceGap;
    private Boolean seniorityMatch;
    private Integer embeddingScore;
    private java.util.List<SemanticMatchDto.SkillScoreDto> skillScores;
    private java.util.List<SemanticMatchDto.RequirementScoreDto> requirementScores;
    private java.util.List<String> strengths;
    private java.util.List<String> weaknesses;
    private String recommendation;
    private java.util.List<String> interviewQuestions;
    private String scoreExplanation;
    private java.util.List<SemanticMatchDto.WarningDto> warnings;
}
